File: tf-python/DataFitting.py
'''
目的：用平面拟合数据点
算法：随机生成100个点，构建线性模型及优化器，训练200次
输入：无
输出：每十次训练输出“当前步 权重W 一次项b”
'''

import tensorflow as tf
import numpy as np

# 使用 NumPy 生成假数据100个
x_data = np.float32(np.random.rand(2, 100))
y_data = np.dot([0.100, 0.200], x_data) + 0.300

# 构造一个线性模型
W = tf.Variable(tf.random_uniform([1, 2]))
b = tf.Variable(tf.zeros([1]))
y = tf.matmul(W, x_data) + b

# 最小化方差
loss = tf.reduce_mean(tf.square(y - y_data))
optimizer = tf.train.GradientDescentOptimizer(0.5)
train = optimizer.minimize(loss)

# 初始化变量
# 已弃用 init = tf.initialize_all_variables()
init = tf.global_variables_initializer()

# 启动图 (graph)
sess = tf.Session()
sess.run(init)

# 拟合平面
for step in range(0, 201):
    sess.run(train)
    if step % 10 == 0:
        print(step, sess.run(W), sess.run(b))

File: tf-python/DataFitting.js
import * as tf from '@tensorflow/tfjs';

const w = tf.variable(tf.randomUniform([1, 2]));
const b = tf.variable(tf.scalar(Math.random()));

const numIterations = 201;
const learningRate = 0.5;
const optimizer = tf.train.sgd(learningRate);

function generateData(numPoints, coeff, sigma = 0.04) {
  return tf.tidy(() => {
    const [_w, _b] = [
      tf.variable(tf.tensor2d(coeff.w, [1, 2])),
      tf.scalar(coeff.b)
    ];

    const xs = tf.variable(tf.randomUniform([2, numPoints]));
    const ys = tf.matMul(_w, xs).add(_b);

    return {xs, ys};
  });
}

function predict(x) {
  return tf.tidy(() => {
    return tf.matMul(w, x).add(b);
  });
}

function loss(prediction, label) {
  const error = prediction.sub(label).square().mean();
  return error;
}

async function train(xs, ys, numIterations) {
  for (let iter = 0; iter < numIterations; iter++) {
    optimizer.minimize(() => {
      const pred = predict(xs);
      return loss(pred, ys);
    });
    if (iter % 10 === 0) {
      console.log(iter, w.dataSync(), b.dataSync());
    }
    await tf.nextFrame();
  }
}

async function learnCoefficients() {
  const trueCoefficients = {w: [0.1, 0.2], b: 0.3};
  // 生成有误差的训练数据
  const trainingData = generateData(100, trueCoefficients);
  // 训练模型
  await train(trainingData.xs, trainingData.ys, numIterations); 
}

learnCoefficients();
